subprojects/instant-execution/src/integTest/groovy/org/gradle/instantexecution/InstantExecutionTaskWiringIntegrationTest.groovy

```diff
@@ -174,4 +178,28 @@ class InstantExecutionTaskWiringIntegrationTest extends AbstractInstantExecution
         instantExecution.assertStateLoaded()
         result.assertTasksSkipped(":producer", ":transformer")
     }
+
+    def "ActionNode serialization failure carries related task information"() {
+        given: "a task with a WorkNodeAction dependency"
+        buildFile << """
+            task run {
+                dependsOn(
+                    { ${TaskDependencyResolveContext.name} resolveContext ->
+                        resolveContext.add(
+                            new ${WorkNodeAction.name}() {
+                                Project getProject() { null }
+                                void run(${NodeExecutionContext.name} executionContext) {}
+                            }
+                        )
+                    } as ${TaskDependencyContainer.name}
+                )
+            }
+        """
+
+        when:
+        instantRun("run")
+
+        then: "failure message should point to related task"
+        outputContains("- task `:run` of type `org.gradle.api.DefaultTask`: cannot serialize object of type 'org.gradle.execution.plan.ActionNode' as these are not supported with instant execution.")
```

Comment:
I'm not sure I'd make this change, because it's somewhat misleading and probably not better than what was already there. Action nodes are just nodes in the graph, just like an artefact transform or task. It doesn't really make sense to report one of the tasks from which the node is reachable through the node graph (rather than the object graph of the task) any more than it would be to report "task 'check' has a problem" when we can't serialise a field of `compileJava`. Also, this is the message we use to mean "will never be supported so you need to change something" which is not the case here. There's nothing the user can do to remove these nodes.